src/helpers/client/FormValue.js

```diff
@@ -0,0 +1,109 @@
+/** Handles the core of form values, including:
+ * - Storing both a current and future value
+ * - Tracking when the user manipulates a future value
+ * - Propagating current values to future values if they
+ * haven't been manipulated by the user
+ * - Making this instance serializable
+ * - Making sure only valid values are saved.
+ * 
+ * All values stored in a FormValue instance should be valid
+ * 
+ * There must be a better way to write that.
+ * 
+ * @class
+ * @param {string} name - the name of the client property
+ * @param {any} [val] - the initial value
+ * @param {function} [makeValid] - returns a valid value no
+ * matter what. Will be given the new value, the current value,
+ * the string 'current' or 'future', and the `FormValue` instance.
+ * 
+ * @returns FormValue
+ * @todo Finish description of FormValue return value
+ */
+class FormValue {
+
+  constructor ( name, val = null, makeValid ) {
+
+    var fVal = this;
+    
+    fVal.name = name;
+    fVal._futureWasChanged = false;  // Handled internally
+    fVal.makeValid = makeValid || function ( newVal ) { return newVal; };
+    
+    fVal.setCurrent( val );
+
+  }  // End FormValue.constructor()
+
+
+  setCurrent ( val ) {
+
+    var fVal = this;
+
+    // Transformer needed for parseInt and such
+    fVal.current = fVal.makeValid( val, fVal.current, 'current', fVal );
```

Comment:
What happens if makeValid cannot transform the value? I suggest we explicitly raise an error. Certainly, wee should not proceed with execution. Also, if makeValid is just casting values I wonder if there's not a better solution. Ordinarily it would be a strongly typed database, but I imagine there are other solutions here.

Reply:
That's the point of `makeValid()`. It *always* returns a valid value no matter what. As for what it's doing to transform the value, it's not just casting it. In some cases it's converting a string to a number, or a string to a boolean value. Other times, it's checking against values that it is dependent on. For example: household size is changed to 5, then number of dependents is changed to 4, then household size is changed to 3. Number of dependents now needs to change to something below 3.

Reply:
Errors in input, as we said, should be dealt with by components. The client object properties should always be valid. Maybe I'm just not seeing some cases, though. Can you give some examples?

Reply:
If it's mostly used for enforcing min/max values, like in your answer, that's pretty reasonable, though we may consider just using an optional `bounds` param. Then we can clamp the value to be within those bounds as you're suggesting.

For eg. string to boolean though, if the valid answers are `yes`/`no` and you write `'bavadf'`, or `'5'`what are we supposed to do with that? What about string to number (probably a more common case) with a bogus input value? I think most of these can be enforced by using the right input type on the front end. Then we're down to if someone were to console in values in which case, I think a validator will do just fine for us since there's no reason for us to try to convert bogus values. In that case, what cases *are* we concerned about?

I guess my concern here is just that it's really hard to identify all the possible edge cases we'd need to convert, but it's relatively simple to reject any except for those that match a very specific format, then let the user re-enter them correctly.

Regardless, we can put it out there and see how it works in action.